Implement label/screenspace text

hobogenized opened the pull request:
- Switch to free license Caladea font
- Change text example to demonstrate difference between label and text
- Enforce that texture for font exists at same root as font data file
- Change 2d font shader to allow for transforms
- Remove references to models

acgaudette commented on shaders/font2d.vert:
```diff
@@ -1,25 +1,23 @@
 void main(void)
 {
-    gl_Position = vec4(inPos, 1.0);
+    gl_Position = model_data.model * vec4(inPos, 1.0);
```
Maybe comment exactly what's going on lines 19-21

acgaudette commented on shaders/font2d.vert:
```diff
@@ -1,25 +1,23 @@
-} shared_data;
+layout(binding = 0) uniform font_ubo {
+  mat4 model;
+} model_data;
```
I might call this `instance` instead of `model_data`

acgaudette commented on shaders/font2d.vert:
```diff
@@ -1,25 +1,23 @@
+  mat4 model;
+} model_data;
 
 layout (location = 0) in vec3 inPos;
```
Consider making this a `vec2` (and updating the requisite code in the renderer). You could still use the transform component with no issues, but wherever you load the actual position data, you could use an `alg::Vec2` instead. Are you using the z for anything?

acgaudette commented on src/components/label.rs:
```diff
@@ -0,0 +1,138 @@
+        self
+    }
+
+    pub fn pixel_scale_factor(&mut self, scale_factor: f32) -> &mut TextBuilder<'a> {
```
May as well add some doc comments on `pixel_scale_factor` and `aspect_scale_factor` explaining what they are.

acgaudette commented on src/components/label.rs:
```diff
@@ -0,0 +1,138 @@
+}
+
+impl components::Component for Manager {
+
```
```suggestion
```

acgaudette commented on src/components/label.rs:
```diff
@@ -0,0 +1,138 @@
+        self.instances.len()
+    }
+
+    #[cfg(debug_assertions)] fn debug_name(&self) -> &str { "Text" }
```
```suggestion
    #[cfg(debug_assertions)] fn debug_name(&self) -> &str { "Label" }
```

acgaudette commented on src/components/label.rs:
```diff
@@ -0,0 +1,138 @@
+    pub(crate) fn update(&mut self, transforms: &transform::Manager) {
+        self.model_matrices.clear();
+        for (entity, _) in &mut self.instances {
+            let font_ubo = render::FontUBO {
```
It's a bit confusing here that the actual model matrix gets put into a font UBO that is then pushed into a vec of "model matrices." Perhaps could rename `model_matrices` to `uniform_data` or `instance_data`.

If you go down this path, it would be a good idea to do it for the text component as well.

acgaudette commented on src/components/mod.rs:
```diff
@@ -37,10 +38,11 @@ pub trait Component {
-    pub draws:      draw::Manager,
-    pub softbodies: softbody::Manager,
-    pub texts:      text::Manager,
+    pub transforms:     transform::Manager,
```
Looks like something weird happened here with the spacing.

acgaudette commented on src/components/text_overlay.rs:
```diff
@@ -0,0 +1,137 @@
+extern crate fnv;
```
Is this file necessary?

acgaudette commented on src/render.rs:
```diff
@@ -831,13 +890,21 @@ impl<'a> Context<'a> {
+        self.device.free_memory(self.text_display.vertex_memory, None);
+        self.device.destroy_buffer(self.text_display.font_ubo_buffer, None);
+        self.device.free_memory(self.text_display.font_ubo_memory, None);
+        /*
```
What's going on here?

acgaudette commented on src/render.rs:
```diff
@@ -3019,9 +3086,16 @@ fn set_image_layout_helper(
+        image_memory_barrier_builder = image_memory_barrier_builder
+            .dst_access_mask(dst_access_mask)
+    }
     let image_memory_barrier = image_memory_barrier_builder
```
```suggestion

    let image_memory_barrier = image_memory_barrier_builder
```

acgaudette committed:
Add font2d compile line to build script

acgaudette committed:
Update text example

Update comments
Reverse 2D animation angle
Slow angle

acgaudette commented on src/components/label.rs:
```diff
@@ -0,0 +1,138 @@
+        self.text.text = text.into();
+        self
+    }
+    
```
```suggestion

```

acgaudette commented on examples/text.rs:
```diff
@@ -46,21 +50,50 @@ impl nmg::Start for Demo {
+            .for_entity(label_1);
+        self.objects.push(label_1);
+
+        /* Point light */
```
Note that your font shaders are currently not computing lighting.

acgaudette commented on examples/text.rs:
```diff
@@ -121,8 +156,23 @@ impl nmg::Update for Demo {
+        // Rotate label
+        components.transforms.set_orientation(
+            self.label_1.unwrap(),
+            alg::Quat::axis_angle_raw(alg::Vec3::up(), -angle)
```
To be honest, I'd prefer it if we only keep the second rotation (line 170) for the example.

acgaudette committed:
Track resources directory for LFS

hobogenized committed:
Add support for new lines
- Modify example text for demonstration

hobogenized committed:
Add font files via git-lfs

acgaudette commented:
Also @hobogenized there was that screen resize scaling bug.

hobogenized committed:
Refactor to use Vec2 for inPos on 2d text vert shader
- Rename model_matrices to instance_data
- Reformat code

hobogenized committed:
Implement label/screenspace text
- Switch to free license Caladea font
- Change text example to demonstrate difference between label and text
- Enforce that texture for font exists at same root as font data file
- Change 2d font shader to allow for transforms
- Remove references to models

acgaudette committed:
Update spacing in mod.rs

acgaudette commented on src/render.rs:
```diff
@@ -3545,21 +3504,41 @@ pub struct TextInstance {
+    index_memory: vd::DeviceMemoryHandle,
+    font_ubo_buffer: vd::BufferHandle,
+    font_ubo_memory: vd::DeviceMemoryHandle,
+    //binding_description: [vd::VertexInputBindingDescription; 1],
```
Do you need these?

acgaudette commented on src/render.rs:
```diff
@@ -3545,21 +3504,41 @@ pub struct TextInstance {
+            ([FontVertex_3d::binding_description()],
+            FontVertex_3d::attribute_descriptions())
+        };
     // Loading shaders
```
```suggestion

    // Load shaders
```

acgaudette commented on src/render.rs:
```diff
@@ -3651,43 +3804,52 @@ fn create_text(
+            index_memory,
+            font_ubo_buffer,
+            font_ubo_memory,
+            //binding_description,
```
Same question here.

acgaudette reviewed (approved):
Looks good. My primary point of feedback is that 2D and 3D text code is not clearly segregated. In one instance we have a generic function, in another we have a bool and conditional checks. I understand that there is a lot of shared functionality, but it is likely better to split it especially if we make further diverging changes in the future. Not too worried right now given that it appears this is not a problem at runtime i.e. the update loop.

hobogenized committed:
Remove unused commented code

hobogenized commented:
Sounds good. We'll try refactoring in the future.
